frame/support/procedural/src/pallet/parse/call.rs

```diff
@@ -0,0 +1,237 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::helper;
+use quote::ToTokens;
+use syn::spanned::Spanned;
+
+/// List of additional token to be used for parsing.
+mod keyword {
+	syn::custom_keyword!(DispatchResultWithPostInfo);
+	syn::custom_keyword!(Call);
+	syn::custom_keyword!(OriginFor);
+	syn::custom_keyword!(weight);
+	syn::custom_keyword!(compact);
+	syn::custom_keyword!(T);
+	syn::custom_keyword!(pallet);
+}
+
+/// Definition of dispatchables typically `impl<T: Config> Pallet<T> { ... }`
+pub struct CallDef {
+	/// The where_clause used.
+	pub where_clause: Option<syn::WhereClause>,
+	/// A set of usage of instance, must be check for consistency with trait.
+	pub instances: Vec<helper::InstanceUsage>,
+	/// The index of call item in pallet module.
+	pub index: usize,
+	/// Information on methods (used for expansion).
+	pub methods: Vec<CallVariantDef>,
+	/// The span of the attribute.
+	pub attr_span: proc_macro2::Span,
+}
+
+/// Definition of dispatchable typically: `#[weight...] fn foo(origin .., param1: ...) -> ..`
+pub struct CallVariantDef {
+	/// Function name.
+	pub fn_: syn::Ident,
+	/// Information on args: `(is_compact, name, type)`
+	pub args: Vec<(bool, syn::Ident, Box<syn::Type>)>,
+	/// Weight formula.
+	pub weight: syn::Expr,
+	/// Docs, used for metadata.
+	pub docs: Vec<syn::Lit>,
+}
+
+/// Attributes for functions in call impl block.
+/// Parse for `#[pallet::weight = expr]`
+pub struct FunctionAttr {
+	weight: syn::Expr,
+}
+
+impl syn::parse::Parse for FunctionAttr {
+	fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+		input.parse::<syn::Token![#]>()?;
```

Comment:
I wonder: isn't there a way to implement this when being parsed as a proper attribute, rather than manually parsing tokens like `#` and `pallet` and `::` one by one? I am not sure though tbh

Reply:
I wondered too, there is indeed `https://substrate.dev/rustdocs/v2.0.0/syn/struct.Attribute.html`, We could implement this `FunctionAttr` to just assert properties on the `path` field and then do the manual parsing of the `token`.

Asserting properties on `path` is not hard, we would check that there is 2 path segments, that they have no arguments and that they have the desired name.
But I though it would take a bit more lines than just checking for `#[pallet::weight`, so I did this way. But now you say it I understand the other way might look cleaner. Anyway I think it is fine to me.

Reply:
whichever you think looks better. 

I think  introducing a custom struct with a parse impl for it that just parses `#[pallet::x]` that parses the prefix and the last `]`, and returns `x` for further parsing, is also a good helper. 

But all in all this is not an important issues. Just some nit ideas.